crates/ty_python_semantic/src/types/protocol_class.rs

```diff
@@ -57,58 +57,69 @@ impl<'db> Deref for ProtocolClassLiteral<'db> {
-#[derive(Debug, PartialEq, Eq, salsa::Update, Default, Clone, Hash)]
-pub(super) struct ProtocolInterface<'db>(BTreeMap<Name, ProtocolMemberData<'db>>);
+#[salsa::interned(debug)]
+pub(super) struct ProtocolInterface<'db> {
+    #[return_ref]
+    _members: BTreeMap<Name, ProtocolMemberData<'db>>,
+}
 
 impl<'db> ProtocolInterface<'db> {
     /// Iterate over the members of this protocol.
-    pub(super) fn members<'a>(&'a self) -> impl ExactSizeIterator<Item = ProtocolMember<'a, 'db>> {
-        self.0.iter().map(|(name, data)| ProtocolMember {
+    pub(super) fn members<'a>(
+        &'a self,
+        db: &'db dyn Db,
+    ) -> impl ExactSizeIterator<Item = ProtocolMember<'a, 'db>> {
+        self._members(db).iter().map(|(name, data)| ProtocolMember {
             name,
             ty: data.ty,
             qualifiers: data.qualifiers,
         })
     }
 
-    pub(super) fn member_by_name<'a>(&self, name: &'a str) -> Option<ProtocolMember<'a, 'db>> {
-        self.0.get(name).map(|data| ProtocolMember {
+    pub(super) fn member_by_name<'a>(
+        self,
+        db: &'db dyn Db,
+        name: &'a str,
+    ) -> Option<ProtocolMember<'a, 'db>> {
+        self._members(db).get(name).map(|data| ProtocolMember {
             name,
             ty: data.ty,
             qualifiers: data.qualifiers,
         })
     }
 
     /// Return `true` if all members of this protocol are fully static.
-    pub(super) fn is_fully_static(&self, db: &'db dyn Db) -> bool {
-        self.members().all(|member| member.ty.is_fully_static(db))
+    pub(super) fn is_fully_static(self, db: &'db dyn Db) -> bool {
+        cached_is_fully_static(db, self)
     }
 
     /// Return `true` if if all members on `self` are also members of `other`.
     ///
     /// TODO: this method should consider the types of the members as well as their names.
-    pub(super) fn is_sub_interface_of(&self, other: &Self) -> bool {
-        self.0
+    pub(super) fn is_sub_interface_of(self, db: &'db dyn Db, other: Self) -> bool {
+        self._members(db)
             .keys()
-            .all(|member_name| other.0.contains_key(member_name))
+            .all(|member_name| other._members(db).contains_key(member_name))
     }
 
     /// Return `true` if any of the members of this protocol type contain any `Todo` types.
-    pub(super) fn contains_todo(&self, db: &'db dyn Db) -> bool {
-        self.members().any(|member| member.ty.contains_todo(db))
+    pub(super) fn contains_todo(self, db: &'db dyn Db) -> bool {
+        self.members(db).any(|member| member.ty.contains_todo(db))
     }
 
     pub(super) fn normalized(self, db: &'db dyn Db) -> Self {
-        Self(
-            self.0
-                .into_iter()
-                .map(|(name, data)| (name, data.normalized(db)))
-                .collect(),
+        Self::new(
+            db,
+            self._members(db)
+                .iter()
+                .map(|(name, data)| (name.clone(), data.clone().normalized(db)))
```

Comment:
Not sure if there is a way to avoid cloning here, since `_members(db)` only gives us shared access to the `BTreeMap`.

Reply:
I think you could change `ProtocolMemberData::normalized()` to take `&self` rather than `self`. It probably wouldn't improve perf, but it would avoid the need for one of the explicit `.clone()` calls, which maybe makes us feel slightly better about life?